experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp

```diff
@@ -71,6 +83,187 @@ LogicalResult HandshakeSpeculationPass::placeUnits(Value ctrlSignal) {
   return success();
 }
 
+// This recursive function traverses the IR and creates a control path
+// by replicating the branches along the way. It stops at commits and
+// connects them to the newly created control path, with value ctrlSignal
+bool HandshakeSpeculationPass::routeBranchControlTraversal(
+    llvm::DenseSet<Operation *> &visited, Value ctrlSignal, Operation *currOp) {
+  // End traversal if already visited
+  if (visited.contains(currOp))
+    return false;
+  visited.insert(currOp);
+
+  bool found_commit = false;
+  if (isa<handshake::SpecCommitOp>(currOp)) {
+    // Connect commit to the correct control signal and end traversal
+    currOp->setOperand(1, ctrlSignal);
+    found_commit = true;
+  } else if (isa<handshake::ConditionalBranchOp>(currOp)) {
+    // Replicate a branch in the control path and use new control signal
+    MLIRContext *ctx = &getContext();
+    OpBuilder builder(ctx);
+
+    builder.setInsertionPointAfterValue(ctrlSignal);
+
+    // The Speculating Branch will discard the control token if the token
+    // in the speculative path is non-speculative.
+    Operation *branchDisc = builder.create<handshake::SpeculatingBranchOp>(
+        currOp->getLoc(), currOp->getResult(0), // specTagOperand
+        currOp->getOperand(0)                   // dataOperand
+    );
+    visited.insert(branchDisc);
+
+    // Connect a conditional branch at the true result of branchDisc
+    Operation *branchCond = builder.create<handshake::ConditionalBranchOp>(
+        branchDisc->getLoc(),
+        branchDisc->getResult(0), // conditionOperand
+        ctrlSignal                // dataOperand
+    );
+    visited.insert(branchCond);
+
+    // Follow the two branch results with a different control signal
+    for (int i = 0; i <= 1; ++i) {
+      for (Operation *succOp : currOp->getResult(i).getUsers()) {
+        Value ctrl = branchCond->getResult(i);
+        bool routed = routeBranchControlTraversal(visited, ctrl, succOp);
+        found_commit = found_commit || routed;
+      }
+    }
+
+    if (!found_commit) {
+      // Remove unused branch signal
+      branchCond->erase();
+      branchDisc->erase();
+    }
+
+  } else {
+    // Continue Traversal
+    for (Value res : currOp->getResults()) {
+      for (Operation *succOp : res.getUsers()) {
+        bool routed = routeBranchControlTraversal(visited, ctrlSignal, succOp);
+        found_commit = found_commit || routed;
+      }
+    }
+  }
+  return found_commit;
+}
+
+LogicalResult HandshakeSpeculationPass::prepareAndPlaceCommits() {
+  // Place commits and connect to the Speculator Commit Control Signal
+  Value commitCtrl = specOp.getCommitCtrl();
+  if (failed(placeUnits<handshake::SpecCommitOp>(commitCtrl)))
+    return failure();
+
+  // Create visited set
+  llvm::DenseSet<Operation *> visited;
+  visited.insert(specOp.getDataOut().getDefiningOp());
```

Comment:
It's always true that `specOp.getDataOut().getDefiningOp() == specOp` right? If so then just use `specOp`